Add `Marketplace` class for Reactified marketplace

andfinally opened the pull request:
Related: 17315-gh-Automattic/woocommerce.com

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

- Adds Marketplace class in Internal/Admin which registers page route.
- Adds `MarketplaceServiceProvider` to make the class available.
- Adds `marketplace` item to `FeaturesController` and enables the feature by default.
- Gets the new `Marketplace` feature in `class-wc-admin-menus.php` if the feature is enabled.
- Adds `Marketplace` menu item to WooCommerce menu.
- Adds components for the `Discover` and `Extensions` tabs.

### Submission Review Guidelines:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes 17315-gh-Automattic/woocommerce.com. This branch is meant to provide the basic routes and pages for a new React-powered marketplace, replacing the current `Extensions` pages.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Check out this branch.
2. Enable the right Node version:

```
nvm use
```

3. From the WooCommerce plugin folder, update the Composer autoload class maps:

```
cd plugins/woocommerce
composer dump-autoload
```

4.  Then run the build:

```
pnpm run --filter=woocommerce build
```

5. Go to WooCommerce in your local environment, and confirm that you see the new menu item `Marketplace` in the WooCommerce menu above `Settings`. (If you're using wp-env the URL is http://localhost:8888/wp-admin/admin.php?page=wc-admin&path=%2Fmarketplace.)
6. Check that you can navigate from there to the `Extensions` tab and back again. The `Marketplace` menu item should be highlighted as the current item while you're on each of these tabs, and the URL should correctly reflect the page and tab. If you visit one of the URLs, the correct tab should load.

### Note

Please ignore these files in the GitHub diff:

```
plugins/woocommerce/tests/e2e-pw/tests/smoke-tests/upload-plugin.spec.js
plugins/woocommerce/changelog/e2e-fix-upload-plugin-spec
```

GitHub is highlighting changes in these files after a rebase on trunk, even though the content in this branch matches that on [trunk](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/tests/e2e-pw/tests/smoke-tests/upload-plugin.spec.js#L118) and the target branch, [feature/marketplace](https://github.com/woocommerce/woocommerce/blob/feature/marketplace/plugins/woocommerce/tests/e2e-pw/tests/smoke-tests/upload-plugin.spec.js#L118).

<!-- End testing instructions -->

### Screenshots

<img width="500" src="https://github.com/woocommerce/woocommerce/assets/1647564/518a581f-d1e7-4deb-b0d5-fb34d7dbf5c1">

andfinally committed:
- Added `Marketplace` class in `Internal/Admin` which registers some page routes.
- Added `MarketplaceServiceProvider` to make the class available.
- Added `marketplace` item to `FeaturesController`.
- Getting the new `Marketplace` feature in `class-wc-admin-menus.php` if the feature is enabled.

andfinally committed:
- Added changelog.

andfinally committed:
Added tsx and scss files for WC Admin-powered Marketplace.

andfinally committed:
Calling `admin_menu` action with priority 45, so Marketplace menu item is positioned above the Settings menu item.
Changed path of Marketplace Discover page to `/marketplace`, as it is the default page.

andfinally committed:
Importing components for Discover, Extensions and Themes marketplace pages. Pushing the pages to the `pages` array if the `marketplace` feature is enabled.
Modified `wpNavMenuClassChange` function to highlight parent menu item with `current` CSS class if the current URL contains a slash.

andfinally commented on plugins/woocommerce-admin/client/layout/controller.js:
```diff
@@ -481,14 +540,21 @@ window.wpNavMenuClassChange = function ( page, url ) {
+	let pageUrl = 'admin.php?page=wc-admin',
+		currentItemsSelector = `li > a[href$="${ pageUrl }"], li > a[href*="${ pageUrl }?"]`;
+
+	if ( '/' !== url ) {
+		pageUrl = 'admin.php?page=wc-admin&path=' + encodeURIComponent( url );
+		const splitUrl = url.split('/');
+		if ( '' === splitUrl[0] ) {
+			splitUrl.splice(0, 1);
+		}
+		currentItemsSelector =
+			splitUrl.length > 1
+				? `li > a[href*="${ encodeURIComponent( splitUrl[0] ) }"]`
+				: `li > a[href*="${ pageUrl }"]`;
+	}
```
The intention behind this change is to highlight the parent menu item when we're viewing one of the page's children.

`url` is the `path` parameter passed in the WC Admin URL. If `url` contains a slash, we split it and use the first non-empty fragment as the selector for the current menu item.

So if we're viewing `wp-admin/admin.php?page=wc-admin&path=%2Fmarketplace%2Fextensions`, we'll highlight the menu item for its parent page, `wp-admin/admin.php?page=wc-admin&path=%2Fmarketplace`.

andfinally replied:
Removed this change: using tabs in a single page component is less problematic.

Commit:
Fix `upload-plugin` spec by updating locator from core profiler (#38914)

* Update locator

* Add changelog

andfinally commented:
@mlaetitia noticed an issue with the menu item link highlighting: all links in Analytics get the `current` class, because their URLs include slashes. I'm putting this back to draft till I can figure out a better solution.

andfinally committed:
Changed routing, so we have one page, `path=/marketplace`, and tabs for Discover and Extensions.

andfinally committed:
Removing registration of unneeded pages from `Marketplace` class.

andfinally committed:
Addressing linter errors.

andfinally committed:
Addressing linter errors.

andfinally committed:
Addressing linter errors.

andfinally committed:
Correcting error in registration of pages introduced in bff51c0e74407df2aa18f467453b0299e92df52a.

andfinally committed:
Fixing linter errors.

kdevnel commented on plugins/woocommerce-admin/client/layout/controller.js:
```diff
@@ -173,6 +176,25 @@ export const getPages = () => {
+		pages.push( {
+			container: Marketplace,
+			layout: {
+				header: false,
```
**Non-review question:** 

Is this a simple way to disable the WC Admin top bar on this page? If so, this is handy!

andfinally replied:
Yes it is! The CSS still leaves space for a header at the top of the page, but we can fill that with our own header.

kdevnel commented on plugins/woocommerce-admin/client/marketplace/components/tabs/tabs.tsx:
```diff
@@ -0,0 +1,99 @@
+			setSelectedTab( query.tab );
+		}
+		// Only run once
+		// eslint-disable-next-line react-hooks/exhaustive-deps
```
**Non review question:**

Just clarifying that we're ignoring because there is no dependency on this `useEffect` hook. You've excluded a dependency, so this runs only a single time?

andfinally replied:
Yes, we only want to run this once, but ESLint demands we list any dependencies, so it throws the warning:

```
React Hook useEffect has a missing dependency: 'setSelectedTab'. Either include it or remove the dependency array
```

But on consideration, it's better not to use an `eslint-disable` comment, so I've added the dependency in in [`e2b61c5` (#38885)](https://github.com/woocommerce/woocommerce/pull/38885/commits/e2b61c5b037653a983f999a57495ae2adcc921d2). Even with it there, this function will only run once.

kdevnel commented on plugins/woocommerce-admin/client/marketplace/index.tsx:
```diff
@@ -0,0 +1,25 @@
+import Content from './components/content/content';
+
+export default function Marketplace() {
+	const [ selectedTab, setSelectedTab ] = useState( 'discover' );
```
I'm not sure if it matters, but when you manually change the URL to one that doesn't exist (such as the themes tab `wp-admin/admin.php?page=wc-admin&path=%2Fmarketplace&tab=themes`), I notice that Discover is the default but isn't highlighted. This is very nitpicky, and I have no preference here. Just mentioning in case it's something you're bothered about.

andfinally replied:
Ah good spot, thanks! I've fixed this in [`e2b61c5` (#38885)](https://github.com/woocommerce/woocommerce/pull/38885/commits/e2b61c5b037653a983f999a57495ae2adcc921d2).

kdevnel reviewed (approved):
This seems to work as described. I'd make a couple of changes to the testing instructions to be safe. I added some questions for general discussion but nothing blocking so :shipit: 

After step 1, I'd add a reminder to `nvm use` from the project root. I forgot that step and it tripped me up.

For step 3, "project root" for me wasn't clear. I ran this command from the top of the WooCommerce repo and got an error. Ran it again from the same folder as step 2 (`plugins/woocommerce`) and it worked from there.

Finally, for step 5, you mention themes. We're not including themes at this point and so this doesn't work (although manually navigating to themes just takes you to DIscover). Should this say "Discover" instead?

andfinally committed:
Added `discover` key for default tab to constants.
Added `setSelectedTab` to dependencies for `useEffect` function that sets the selected tab based on the URL. Simplified the logic in this function – we can rely on the `path` always being `MARKETPLACE_PATH`. If tab in URL is not a recognised tab, set the default tab as selected.

andfinally committed:
Added comment to empty SCSS file to silence stylelint `no-empty-source` error.
